Add `CategoricalMADE`

jnsbck opened the pull request:
## What does this implement/fix? Explain your changes
This implements a `CategoricalMADE` to generelize MNLE to multiple discrete dimensions addressing  #1112.
Essentially adapts `nflows`'s MixtureofGaussiansMADE to autoregressively model categorical distributions.

## Does this close any currently open issues?

Fixes #1112


## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask. We're here to
help! This is simply a reminder of what we are going to look for before merging
your code.

- [ ] I have read and understood the [contribution
  guidelines](https://github.com/sbi-dev/sbi/blob/main/CONTRIBUTING.md)
- [ ] I agree with re-licensing my contribution from AGPLv3 to Apache-2.0.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have reported how long the new tests run and potentially marked them
  with `pytest.mark.slow`.
- [ ] New and existing unit tests pass locally with my changes
- [ ] I performed linting and formatting as described in the [contribution
  guidelines](https://github.com/sbi-dev/sbi/blob/main/CONTRIBUTING.md)
- [ ] I rebased on `main` (or there are no conflicts with `main`)
- [ ] For reviewer: The continuous deployment (CD) workflow are passing.

jnsbck commented:
Hey @janfb,
would very much appreciate your input at this stage:

Currently the PR adds the [`CategoricalMADE`](https://github.com/jnsbck/sbi/blob/bcc75dbdf4ab39533b69d9466dc1e4737468d4cf/sbi/neural_nets/estimators/categorical_net.py#L17) and builder [`build_autoregressive_categoricalestimator`](https://github.com/jnsbck/sbi/blob/bcc75dbdf4ab39533b69d9466dc1e4737468d4cf/sbi/neural_nets/net_builders/categorial.py#L69) + some minor modifications to `build_mnle` and `MixedDensityEstimator`. This enables multiple discrete variables with different numbers of classes via `trainer = MNLE(density_estimator=lambda x,y: build_mnle(y,x,categorical_model="made"))` Note that for some reason x and y have to be flipped for mnle.

As far as I can tell all functionalities of `CategoricalMADE` work for both 1D and ND inputs and running the `Example_01_DecisionMakingModel.ipynb` with the CatMADE matches the ground truth 
![image](https://github.com/user-attachments/assets/95def2cd-f77c-4536-b3c4-5aa60c672681)

The question now is: How should I verify this works? / Which tests should I add/modify? Do you have an idea for a good toy example with several discrete variables that I could use?

I have cooked up a toy simulator, for which I am getting good posteriors using SNPE, but for some reason MNLE raises a `RuntimeError: probability tensor contains either 'inf', 'nan' or element < 0` Even for the unmodified MNLE. Any ideas why this could be?

This is the simulator
```python
def toy_simulator(theta: torch.Tensor, centers: list[torch.Tensor]) -> torch.Tensor:
    batch_size, n_dimensions = theta.shape
    assert len(centers) == n_dimensions, "Number of center sets must match theta dimensions"
    
    # Calculate discrete classes by assiging to the closest center
    x_disc = torch.stack([
        torch.argmin(torch.abs(centers[i].unsqueeze(1) - theta[:, i].unsqueeze(0)), dim=0)
        for i in range(n_dimensions)
    ], dim=1)

    closest_centers = torch.stack([centers[i][x_disc[:, i]] for i in range(n_dimensions)], dim=1)
    # Add Gaussian noise to assigned class centers
    std = 0.4
    x_cont = closest_centers + std * torch.randn_like(closest_centers)
       
    return torch.cat([x_cont, x_disc], dim=1)
```

The setup:
```python
torch.random.manual_seed(0)
centers = [
    torch.tensor([-0.5, 0.5]),
    # torch.tensor([-1.0, 0.0, 1.0]),
]

prior = BoxUniform(low=torch.tensor([-2.0]*len(centers)), high=torch.tensor([2.0]*len(centers)))
theta = prior.sample((20000,))
x = toy_simulator(theta, centers)

theta_o = prior.sample((1,))
x_o = toy_simulator(theta_o, centers)
```

NPE:
```python
trainer = SNPE()
estimator = trainer.append_simulations(theta=theta, x=x).train(training_batch_size=1000)

snpe_posterior = trainer.build_posterior(prior=prior)
posterior_samples = snpe_posterior.sample((2000,), x=x_o)
pairplot(posterior_samples, limits=[[-2, 2], [-2, 2]], figsize=(5, 5), points=theta_o)
```

and the equivalent MNLE:
```python
trainer = MNLE()
estimator = trainer.append_simulations(theta=theta, x=x).train(training_batch_size=1000)

mnle_posterior = trainer.build_posterior(prior=prior)
mnle_samples = mnle_posterior.sample((10000,), x=x_o)
pairplot(mnle_samples, limits=[[-2, 2], [-2, 2]], figsize=(5, 5), points=theta_o)
```

Thanks for your feedback.

jnsbck commented:
Hey @janfb,
you might have missed this, but I would be happy for feedback :)

janfb commented on sbi/neural_nets/estimators/mixed_density_estimator.py:
```diff
@@ -80,8 +80,10 @@ def sample(
             )
-            # Trailing `1` because `Categorical` has event_shape `()`.
-            discrete_samples = discrete_samples.reshape(num_samples * batch_dim, 1)
+            num_variables = self.discrete_net.net.num_variables
+            discrete_samples = discrete_samples.reshape(
+                num_samples * batch_dim, num_variables
+            )
```
👍

janfb commented on sbi/neural_nets/estimators/mixed_density_estimator.py:
```diff
@@ -145,7 +147,8 @@ def log_prob(self, input: Tensor, condition: Tensor) -> Tensor:
         )
 
-        cont_input, disc_input = _separate_input(input)
+        num_disc = self.discrete_net.net.num_variables
```
```suggestion
        num_discrete_variables = self.discrete_net.net.num_variables
```

janfb commented on sbi/neural_nets/estimators/mixed_density_estimator.py:
```diff
@@ -145,7 +147,8 @@ def log_prob(self, input: Tensor, condition: Tensor) -> Tensor:
 
-        cont_input, disc_input = _separate_input(input)
+        num_disc = self.discrete_net.net.num_variables
+        cont_input, disc_input = _separate_input(input, num_discrete_columns=num_disc)
```
```suggestion
        cont_input, disc_input = _separate_input(input, num_discrete_variables)
```

janfb commented on sbi/neural_nets/net_builders/categorial.py:
```diff
@@ -61,3 +64,60 @@ def build_categoricalmassestimator(
+    z_score_y: Optional[str] = "independent",
+    num_hidden: int = 20,
+    num_layers: int = 2,
+    categories: Optional[Tensor] = None,
```
add to docstring. 

I suggest renaming this to `num_categories`. and I like the approach with the warning that if it is `None ` it will be inferred from the data.

janfb commented on sbi/neural_nets/net_builders/categorial.py:
```diff
@@ -61,3 +64,60 @@ def build_categoricalmassestimator(
+            standardizing_net(batch_y, structured_y), embedding_net
+        )
+
+    batch_x_discrete = batch_x[:, _is_discrete(batch_x)]
+    inferred_categories = tensor([unique(col).numel() for col in batch_x_discrete.T])
```
these two lines should be executed only `if categories is None`.

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -4,13 +4,116 @@
+        if use_residual_blocks and random_mask:
+            raise ValueError("Residual blocks can't be used with random masks.")
+
+        self.num_variables = len(categories)
+        self.num_categories = int(max(categories))
+        self.categories = categories
```
I am confused by the notion of these three variables. 

what's the difference between variables and categories here?

janfb replied:
what type does `categories` have?

jnsbck replied:
categories lists the number of categories for each categorical variable, i.e. `x_1 = ["heads", "tails"]; x_2 = [1,2,3,4,5,6]` would result in `categories = [2,6]`. I agree the naming might not be ideal. Any suggestions? Also w.r.t. `num_categories`, `num_variables`? Could rename to just `variables` or sth.

janfb commented on sbi/neural_nets/net_builders/mnle.py:
```diff
@@ -125,13 +131,14 @@ def build_mnle(
     )
     # Separate continuous and discrete data.
-    cont_x, disc_x = _separate_input(batch_x)
+    num_disc = int(torch.sum(_is_discrete(batch_x)))
```
please add `num_categorical_columns` as input arg to this function to enable users to pass this number. Inferring it from `batch_x` should be the fallback with warning.

janfb commented on sbi/neural_nets/net_builders/mnle.py:
```diff
@@ -144,15 +151,31 @@ def build_mnle(
+            z_score_y="none",  # y-embedding net already z-scores.
+            num_hidden=hidden_features,
+            num_layers=hidden_layers,
+            embedding_net=embedding_net,
```
pass `num_categories` here as well.

janfb commented on sbi/neural_nets/net_builders/mnle.py:
```diff
@@ -144,15 +151,31 @@ def build_mnle(
+            num_layers=hidden_layers,
+            embedding_net=embedding_net,
+        )
+    elif categorical_model == "mlp":
+        assert num_disc == 1, "MLP only supports 1D input."
+        discrete_net = build_categoricalmassestimator(
+            disc_x,
+            batch_y,
+            z_score_x="none",  # discrete data should not be z-scored.
+            z_score_y="none",  # y-embedding net already z-scores.
+            num_hidden=hidden_features,
+            num_layers=hidden_layers,
+            embedding_net=embedding_net,
+        )
```
more generally, isn't the MLP a special case of the MADE? can't we absorb them into one class?

jnsbck replied:
Comment: Check if testcase is identical, if yes -> rm MLP

jnsbck replied:
This seems to be the case for the examples in the tutorial notebook
![image](https://github.com/user-attachments/assets/f58eb477-0595-4e2a-8079-94cc6425af8d)

![image](https://github.com/user-attachments/assets/939cfdfd-711c-4203-8f5e-003f63c43cbe)

jnsbck replied:
c2st between true and MADE MNLE posterior: 0.538
c2st between true and MLP MNLE posterior: 0.5730000000000001
c2st between MADE MNLE and MLP MNLE posterior: 0.5734999999999999

janfb replied:
that's great! 👍

janfb commented on sbi/made_mnle.ipynb:
this should eventually be integrated with the MNLE tutorial in `12_iid_data_and_permutation_invariant_embeddings.ipynb`

jnsbck replied:
change "mlp" to "made" and comment that several variables with different num_categories are supported.

janfb reviewed (commented):
thanks a lot for tackling this @jnsbck! 👏 

Please find below some comments and questions. 
There might be some misunderstanding about `variables` and `categories` on my side. We can have a call if that's more efficient than commenting here.

jnsbck commented:
Cool, thanks for all the feedback! A quick call would be great, also to discuss suitable tests for this. Will reach out via email and tackle the straight forward things in the meantime.

jnsbck commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -4,13 +4,116 @@
+
+    # outputs (batch_size, num_variables, num_categories)
+    def log_prob(self, inputs, context=None):
+        outputs = self.forward(inputs, context=context)
```
are these shapes correct?

jnsbck commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -4,13 +4,116 @@
+
+        return samples.reshape(*sample_shape, self.num_variables)
+
+    def _initialize(self):
+        pass
```
Yes, this is part of nflows

jnsbck replied:
however, cannot raise not implemented, since this is run on init of MADE.

janfb replied:
Unless I am missing something the `_initialize()` is needed only in `MixtureOfGaussiansMADE(MADE):`, not in `MADE`, so it's not needed here?

jnsbck commented:
After discussion with  @janfb I will:
1. adapt the simulator of `Example_01_DecisionMakingModel.ipynb` to multiple discrete variables.
2. Get this to run for 1D and ND
3. fix remaining comments/issues
4. possibly refactor (fold 1D into ND code).

@janfb could you still check tho what is up with the simulator above? Do you have a hunch why the SNPE and MNLE posteriors different?

jnsbck commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -4,13 +4,116 @@
+
+        with torch.no_grad():
+            samples = torch.zeros(num_samples, self.num_variables)
+            for variable in range(self.num_variables):
```
Samples need to be generated autoregressively, so I dont think so.

jnsbck commented:
I did a bit more work on this PR, all current tests should be passing and I have swapped out all the legacy `CategoricalNet` code for the `CategoricalMADE`. See changes and comments above (please close if no longer relevant). 
A few things remain:
- [ ] Add a bit more docs and comments
- [ ] add test cases
- [ ] adapt the tutorial to 2d? (Can just add another beta distribution to the prior)
- [ ] make sure it runs for ND.

This last thing has been haunting me in my sleep, as I cannot figure out what is wrong. Maybe you have an idea of what could be causing this.
For 1D it works, but for ND it always gets the first discrete dim wrong, i.e. yields the prior (all other dims are correct, see image). I am not sure if the conditioning for the first dimension is broken somehow, but I am not able to pin down where this would be happening in my code.

![image](https://github.com/user-attachments/assets/1877f077-c0da-4b7f-82c3-366a818fd77f)

Help would be much appreciated. @janfb

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,93 +23,134 @@ class CategoricalNet(nn.Module):
+        with torch.no_grad():
+            samples = torch.zeros(num_samples, batch_dim, self.num_variables)
+            print(samples.shape, context.shape)
+            for i in range(self.num_variables):
+                outputs = self.forward(samples, context)
+                outputs = outputs.reshape(*samples.shape, self.num_categories)
+                ps = self.compute_probs(outputs)
+                samples[:, :, i] = Categorical(probs=ps[:, :, i]).sample()
```
same question as above: these samples are internally autoregressive, right? So each discrete variable is sampled given the upstream discrete variables?

janfb replied:
I am just confused because I would have expected that we for each iteration we need pass the so far sampled discrete samples as `context`; but this seems to be happening implicitly in the MADE?

janfb replied:
Now I see it: in line 148 you are updating `samples` with the new samples of the current `i`. It probably boils down to the same thing, but you could also update all sofar sampled `samples`, i.e., 

```
amples[:, :, :(i+1)] = Categorical(probs=ps[:, :, :(i+1)]).sample()
```
?

jnsbck replied:
I dont think it will make a difference whether the prev dims are updated since the dims are autoregressive, but I can make that change

janfb replied:
no need, it was just for my understanding. thanks

janfb commented on sbi/neural_nets/net_builders/mnle.py:
```diff
@@ -102,6 +105,10 @@ def build_mnle(
             as z_score_x.
         flow_model: type of flow model to use for the continuous part of the
             data.
+        categorical_model: type of categorical net to use for the discrete part of
+            the data. Can be "made" or "mlp".
```
add a note here that MLP works only for 1D / one column.

jnsbck replied:
good catch! forgot to remove this, since the MLP was removed anyways!

janfb reviewed (commented):
thanks for the update! 
Made another round of comments. Happy to have another call to sort them out.

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,93 +23,134 @@ class CategoricalNet(nn.Module):
-            condition: batch of parameters for prediction.
+        with torch.no_grad():
+            samples = torch.zeros(num_samples, batch_dim, self.num_variables)
+            print(samples.shape, context.shape)
```
debugging leftover?

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,93 +23,134 @@ class CategoricalNet(nn.Module):
-            sample_shape: number of samples to obtain.
-            condition: batch of parameters for prediction.
+        with torch.no_grad():
+            samples = torch.zeros(num_samples, batch_dim, self.num_variables)
```
should this be initialized with uniform `torch.rand`?

janfb reviewed (commented):
had another look and made two suggestion which could be a reason for the missing first dim fit.

jnsbck commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,93 +23,134 @@ class CategoricalNet(nn.Module):
-        """Returns samples from categorical random variable with probs predicted from
-        the neural net.
+        outputs = self.forward(inputs, context=context)
+        outputs = outputs.reshape(*inputs.shape, self.num_categories)
+        ps = self.compute_probs(outputs)
+
+        # categorical log prob
+        log_prob = torch.log(ps.gather(-1, inputs.unsqueeze(-1).long()))
+        log_prob = log_prob.squeeze(-1).sum(dim=-1)
```
The `gather` is essentially doing the job of the `Categorical` here. If I am not mistaken, the two lines below should be equivalent. Can use the Categorical if you prefer.

```python
log_prob = Categorical(logits=outputs).log_prob(input)
# equivalent
log_prob = F.log_softmax(outputs, dim=-1).gather(-1, input.unsqueeze(-1).long())
```

jnsbck commented:
Thanks for all the input <3, looking into the remaining ones over the coming days hopefully

jnsbck commented:
Turns out, since the posterior is an `MCMCPosterior`, only `log_prob` is used for sampling and not `sample`. This bug still eludes me. 

I have spent some time today and been able to rule a lot of things out (i.e. posterior, sampling, MNLE related things...), which is great, but nonetheless I am still stuck. I have been able to reduce it to the following example of just training a `CategoricalMADE`, which means that whatever is causing the weird behaviour can be found here. I am sure I am completely missing something probably obvious. I hope the code below makes it easier to identify.
@dgedon , @janfb .

```python
#... snle tutorial (incl in this PR)

from sbi.neural_nets.estimators.categorical_net import CategoricalMADE

# Define independent prior.
prior = MultipleIndependent(
    [
        Gamma(torch.tensor([1.0]), torch.tensor([0.5])),
        Beta(torch.tensor([2.0]), torch.tensor([2.0])),
        Beta(torch.tensor([2.0]), torch.tensor([2.0])),
        # Beta(torch.tensor([2.0]), torch.tensor([2.0])),
    ],
    validate_args=False,
)

torch.manual_seed(42)
theta_o = prior.sample((1,))

# Training data
num_simulations = 10000
batch_size = 1000
num_epochs = 100
theta = prior.sample((num_simulations,))
x = mixed_simulator(theta)

# only pred disc dimensions
x = x[:, 1:]

made = CategoricalMADE(
    num_categories=torch.ones(x.shape[1], dtype=torch.int32)*2,
    hidden_features=20,
    context_features=theta.shape[1],
)

in_batches = lambda x: x.reshape(num_simulations // batch_size, batch_size, -1)
optimizer = torch.optim.Adam(made.parameters(), lr=5e-4)
for i in range(num_epochs):
    print(f"\repoch {i+1} / {num_epochs}", end="")
    for theta_batch, x_batch in zip(in_batches(theta), in_batches(x)):
        optimizer.zero_grad()
        loss = -made.log_prob(x_batch, theta_batch).mean()
        loss.backward()
        optimizer.step()

p_true_disc = theta_o[0, 1:]
num_disc = x.shape[1]

# compute marginal probabilities p(x)
choices = torch.arange(2**num_disc).unsqueeze(-1).bitwise_and(2**torch.arange(num_disc)).ne(0).unsqueeze(1)
p_est_disc = torch.zeros(num_disc)
for i in range(num_disc):
    ways_of_choosing_i = choices[torch.any(choices[:, :, i], dim=-1)].float()
    log_prob = made.log_prob(ways_of_choosing_i, theta_o)
    p_est_disc[i] = torch.exp(log_prob).sum().detach()

print("\n")
print(f"true: {p_true_disc}")
print(f"est: {p_est_disc}") # <-- dim=0 incorrect for dim_disc > 1 
```

dgedon commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,93 +23,118 @@ class CategoricalNet(nn.Module):
         """
+        outputs = self.forward(input, condition=condition)
+        outputs = outputs.reshape(*input.shape, self.num_categories)
+        log_prob = Categorical(logits=outputs).log_prob(input).sum(dim=-1)
```
`Categorical(logits=outputs).log_prob(input)` this one gives for:
- one input dimension: varying values for all batches
- multiple input dimensions: the same log_prob for all batches in the first dimension. For all further dimensions, it gives reasonable values. 

This indicates to me that the first dimension in the >1D case is not learnt. But I do not yet understand why this is happening.

dgedon commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,93 +23,118 @@ class CategoricalNet(nn.Module):
+            Log-probabilities of shape `(batch_size, num_variables, num_categories)`.
         """
+        outputs = self.forward(input, condition=condition)
+        outputs = outputs.reshape(*input.shape, self.num_categories)
```
okay, follow up from from the comment one line below:
Here I see for the case of >1D we have that outputs[:,0,:] gives constant values over all batches. Thus, this prevents learning. Why is this? unclear yet

dgedon replied:
It must be in the forward or the initialisation of the MADE

dgedon commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,93 +23,118 @@ class CategoricalNet(nn.Module):
-        # `event_shape=()` but our data usually has an event_shape of `(1,)`.
-        return Categorical(probs=ps).log_prob(input.squeeze(dim=-1))
+        embedded_condition = self.embedding_net.forward(condition)
+        out = super().forward(input, context=embedded_condition)
```
out has constant values for out[:,0] and out[:,1] indicating that the first dimension cannot be learnt. So I assume that MADE is not correctly initialised. But I can't see yet why

dgedon reviewed (commented):
I think it might be the initialisation of the pytorch native MADE. The first dimension in the >1D discrete variables case has constant output over all batches which prohibits any learning. But I couldn't figure out why this happens. Happy to chat though.

janfb commented:
Thanks for the updates @jnsbck ! 


Good catch @dgedon ! I looked into this a big and noticed that actually the first two dimensions in the output of the forward pass remain constant. As a consequence, the log_probs stay constant across the batch as well, and no learning happens. 

I dig into into the `nflows` MADE code to find out why this happens, but I could not find a solution either. I tried changing the initialization e.g., make it a broader uniform to induce more variability in the first autoregressive dimension (which is only influenced by the `bias` term). 
I could not find out why the second dimension remains constant as well. 

I noticed that with only two input features (`autoregressive_features`), the degrees in the MADE are just `[1, 1, 1, ...]`, because 

```
                max_ = max(1, autoregressive_features - 1)
                min_ = min(1, autoregressive_features - 1)
                out_degrees = torch.arange(out_features) % max_ + min_
```
so `min_=max_=1`. 

changing this to 

```
out_degrees = torch.arange(out_features) % autoregressive_features + 1
```

fixes this to be `[1, 2, 1, 2, ...]`, but it does not fix the problem. 

I also noticed that the problem seems to be located in the `final_layer` of the MADE, e.g., 

```
    def forward(self, inputs, context=None):
        temps = self.initial_layer(inputs)
        if context is not None:
            temps += self.context_layer(context)
        for block in self.blocks:
            temps = block(temps, context)
        outputs = self.final_layer(temps)
        return outputs
```
here, all is fine (no constant values over the batch) until the final line. 

I tried to debug the `MaskedLinear` linear that defines the final layer, but could not find the problem. 

Maybe one of you @dgedon or @jnsbck give it a try as well?

dgedon commented:
Thanks @janfb!
> I dig into into the nflows MADE code to find out why this happens

This assumes that the nflows MADE implementation does not work for d>1 dimensions, right? And actually they do not have a test case for MADE in their repository see [here](https://github.com/bayesiains/nflows/tree/master/tests/flows), so this might be reasonable to look into with more care.

jnsbck commented:
Thanks a ton to both of you! I also did a bit more digging, but apart from what @janfb also found, I have nothing conclusive yet.

janfb commented:
another option would be switching to `zuko` instead, e.g., use their

https://github.com/probabilists/zuko/blob/master/zuko/flows/autoregressive.py

instead of `nflows` MADE. They use a `MaskedMLP` as a conditioning net. I could not figure out the shape handling yet though.

gmoss13 commented:
Hi all, had a look at this repo and I found the problem. the `nflows` implementation of MADE doesn't take the context into account correctly. The input to the "first" dimension in MADE gets completely masked out, which is fine for unconditional MADE as the first dimension is generated unconditionally. But as it is currently implemented, `nflows` masks out the input to the last layer, which includes information from the context. I think the easiest fix for this, which I will also put up as a PR in `nflows`, is to simply add a dummy first variable in the MADE implementation in `nflows`, as follows:

```python
class MADE(nn.Module):
    """Implementation of MADE.

    It can use either feedforward blocks or residual blocks (default is residual).
    Optionally, it can use batch norm or dropout within blocks (default is no).
    """

    def __init__(
        self,
        features,
        hidden_features,
        context_features=None,
        num_blocks=2,
        output_multiplier=1,
        use_residual_blocks=True,
        random_mask=False,
        activation=F.relu,
        dropout_probability=0.0,
        use_batch_norm=False,
    ):
        if use_residual_blocks and random_mask:
            raise ValueError("Residual blocks can't be used with random masks.")
        super().__init__()
        self.output_multiplier = output_multiplier
        # Initial layer.
        self.initial_layer = MaskedLinear(
            in_degrees=_get_input_degrees(features+1),
            out_features=hidden_features,
            autoregressive_features=features+1,
            random_mask=random_mask,
            is_output=False,
        )
        

        if context_features is not None:
            self.context_layer_in = nn.Linear(context_features, hidden_features)


        # Residual blocks.
        blocks = []
        if use_residual_blocks:
            block_constructor = MaskedResidualBlock
        else:
            block_constructor = MaskedFeedforwardBlock
        prev_out_degrees = self.initial_layer.degrees
        for _ in range(num_blocks):
            blocks.append(
                block_constructor(
                    in_degrees=prev_out_degrees,
                    autoregressive_features=features+1,
                    context_features=context_features,
                    random_mask=random_mask,
                    activation=activation,
                    dropout_probability=dropout_probability,
                    use_batch_norm=use_batch_norm,
                    zero_initialization=True,
                )
            )
            prev_out_degrees = blocks[-1].degrees
        self.blocks = nn.ModuleList(blocks)

        # Final layer.
        self.final_layer = MaskedLinear(
            in_degrees=prev_out_degrees,
            out_features=(features+1) * output_multiplier,
            autoregressive_features=(features+1),
            random_mask=random_mask,
            is_output=True,
        )

    def forward(self, inputs, context=None):
        dummy_input = torch.zeros((inputs.shape[:-1]+(1,))) # add dummy input to ensure first real dimension of `inputs` is still conditioned on `context`.
        concat_input = torch.cat((dummy_input,inputs),dim=-1)
        temps = self.initial_layer(concat_input)
        if context is not None:
            temps += self.context_layer_in(context)
        for block in self.blocks:
            temps = block(temps, context)
        outputs = self.final_layer(temps)
        return outputs[...,self.output_multiplier:] # remove dummy input
``` 


As far as I understand it, this is a bug in conditional MADE as a whole, unrelated to whether we are estimating categorical or continuous distributions, so for what it's worth @jnsbck I don't think you did anything wrong :)

janfb commented:
@jnsbck this is now fixed by a patch within `sbi`, see https://github.com/sbi-dev/sbi/pull/1398

so you can proceed finishing this PR 🚀

jnsbck commented:
Amazing!

jnsbck commented:
I did it :)
Ready for final send off from my side. Tests passing on my end. Added a comment about option to run with >1D discrete conditions and changed all MNLE tests to pass for discrete_dim  ==2.

Hoping this can be merged if tests pass :)

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,108 +23,134 @@ class CategoricalNet(nn.Module):
-        num_hidden: int = 20,
-        num_layers: int = 2,
-        embedding_net: Optional[nn.Module] = None,
+        num_categories: Tensor,  # Tensor[int]
```
why the comment? is it read by pyright? or can we explicitly use `Tensor[int]` as type hint?

jnsbck replied:
Can only typehint `Tensor` not `Tensor[int]`.

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,108 +23,134 @@ class CategoricalNet(nn.Module):
+                max(categories) defines the number of output units, i.e., the largest
+                number of categories. Can handle mutliple variables with differing
+                numbers of choices.
             num_hidden: number of hidden units per layer.
             num_layers: number of hidden layers.
             embedding_net: emebedding net for input.
```
please adapt docstring.

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,108 +23,134 @@ class CategoricalNet(nn.Module):
-        for _ in range(num_layers):
-            self.hidden_layers.append(nn.Linear(num_hidden, num_hidden))
+        super().__init__(
+            self.num_variables,
+            hidden_features,
```
can we pass with `kwargs` here as well for better readability?

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,108 +23,134 @@ class CategoricalNet(nn.Module):
-        Returns:
-            Tensor: Samples with shape (num_samples, 1)
-        """
+    def sample(self, sample_shape, context=None):
```
`sample_shape: torch.Size,`?

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,108 +23,134 @@ class CategoricalNet(nn.Module):
-        """
+    def sample(self, sample_shape, context=None):
+        # Ensure sample_shape is a tuple
+        if isinstance(sample_shape, int):
+            sample_shape = (sample_shape,)
```
I think we should catch this with type hinting and it will raise an error otherwise.

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,108 +23,134 @@ class CategoricalNet(nn.Module):
+        # Ensure sample_shape is a tuple
+        if isinstance(sample_shape, int):
+            sample_shape = (sample_shape,)
+        sample_shape = torch.Size(sample_shape)
```
not needed then.

janfb commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -21,108 +23,134 @@ class CategoricalNet(nn.Module):
 
-class CategoricalMassEstimator(ConditionalDensityEstimator):
-    """Conditional density (mass) estimation for a categorical random variable.
+        with torch.no_grad():
```
please add a short comment here on how the sampling works.

janfb reviewed (commented):
This is converging now! 🎉 

I just added a couple of minor comments.

jnsbck commented:
Done :) Thanks for the final check up!
